### PR TITLE
K8s: Add Aggregation to Backend Service

### PR DESCRIPTION
**Pull request by toddtreece**
**What is this feature?**

This adds [kube-aggregator](https://github.com/kubernetes/kube-aggregator/) in front of the existing API server when the `kubernetesAggregator` feature toggle is enabled.

**Why do we need this feature?**

`kube-aggregator` acts as an API gateway to connect multiple API servers.

**Who is this feature for?**

@grafana/cloud-app-platform-squad 

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/services/ngalert/schedule/registry_test.go [hunk 1]
```diff
@@ -239,7 +239,7 @@ func TestSchedulableAlertRulesRegistry(t *testing.T) {
 	assert.Len(t, rules, 0)
 	assert.Len(t, folders, 0)
 
-	expectedFolders := map[models.FolderKey]string{models.FolderKey{OrgID: 1, UID: "test-uid"}: "test-title"}
+	expectedFolders := map[models.FolderKey]string{{OrgID: 1, UID: "test-uid"}: "test-title"}
```
Comment (ryantxu) [inline pkg/services/ngalert/schedule/registry_test.go:242]: this seems unrelated...

Comment (toddtreece) [reply pkg/services/ngalert/schedule/registry_test.go:242]: 👍 it looks like this came from [021d2cc](https://github.com/grafana/grafana/pull/81591/commits/021d2cc7e52403c5693bbf21d3fa3e2b0e90d374). i'll revert

**Comment by toddtreece** [comment]
/deploy-to-hg --enterprise-ref toddtreece/standalone-aggregation

**Comment by toddtreece** [comment]
/deploy-to-hg --enterprise-ref toddtreece/standalone-aggregation

**Comment by toddtreece** [comment]
/deploy-to-hg --enterprise-ref toddtreece/standalone-aggregation

File: pkg/services/apiserver/service.go [hunk 2]
```diff
@@ -292,15 +298,54 @@ func (s *service) start(ctx context.Context) error {
...
+		runningServer, err = s.startAggregator(transport, serverConfig, server)
+		if err != nil {
+			return err
+		}
+	} else {
+		runningServer, err = s.startCoreServer(transport, serverConfig, server)
+		if err != nil {
+			return err
+		}
+	}
```
Comment (toddtreece) [inline pkg/services/apiserver/service.go:326]: @ryantxu something like this?

File: pkg/services/apiserver/endpoints/responsewriter/responsewriter.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+	if err := ra.buffered.Writer.Flush(); err != nil {
+		klog.Error("Error flushing response buffer: ", "error", err)
+	}
+}
+
+// Response returns the [http.Response] generated by the [http.Handler].
+func (ra *ResponseAdapter) Response() *http.Response {
+	// make sure to set the status code to 200 if the request is a watch
+	// this is to ensure that client-go uses a stramwatcher:
```
Comment (charandas) [inline pkg/services/apiserver/endpoints/responsewriter/responsewriter.go:82]:
```suggestion
	// this is to ensure that client-go uses a streamwatcher:
```

**Review by charandas — CHANGES_REQUESTED** [review]
Just a nit in a comment string.

File: pkg/services/apiserver/options/options.go [hunk 4]
```diff
@@ -75,6 +75,9 @@ func (o *Options) Validate() []error {
 func (o *Options) ApplyTo(serverConfig *genericapiserver.RecommendedConfig) error {
 	serverConfig.AggregatedDiscoveryGroupManager = aggregated.NewResourceManager("apis")
 
+	// avoid picking up an in-cluster service account token
+	o.RecommendedOptions.Authentication.SkipInClusterLookup = true
```
Comment (toddtreece) [inline pkg/services/apiserver/options/options.go:79]: @charandas i think this might avoid the in cluster issue. thoughts?

File: pkg/services/apiserver/auth/authorizer/provider.go [hunk 5]
```diff
@@ -20,6 +21,7 @@ type GrafanaAuthorizer struct {
 
 func NewGrafanaAuthorizer(cfg *setting.Cfg, orgService orgsvc.Service) *GrafanaAuthorizer {
 	authorizers := []authorizer.Authorizer{
+		authorizerfactory.NewPrivilegedGroups("system:masters"),
```
Comment (toddtreece) [inline pkg/services/apiserver/auth/authorizer/provider.go:24]: any concerns if we always allow requests from the loopback's `system:masters` group?

File: pkg/services/apiserver/auth/authorizer/provider.go [hunk 6]
```diff
@@ -21,6 +22,7 @@ type GrafanaAuthorizer struct {
 func NewGrafanaAuthorizer(cfg *setting.Cfg, orgService orgsvc.Service) *GrafanaAuthorizer {
 	authorizers := []authorizer.Authorizer{
 		&impersonationAuthorizer{},
+		authorizerfactory.NewPrivilegedGroups("system:masters"),
```
Comment (toddtreece) [inline pkg/services/apiserver/auth/authorizer/provider.go:25]: any concerns if we always allow requests from the loopback's system:masters group?

Comment (ryantxu) [reply pkg/services/apiserver/auth/authorizer/provider.go:25]: what does it take to get in the "system:masters" group?  with impersonation disabled, that seems pretty safe

**Comment by toddtreece** [comment]
@ryantxu i added docs and an example based on what @charandas wrote for the standalone aggregator command. remote seems to work as expected:

```
export KUBECONFIG=$PWD/data/grafana-apiserver/grafana.kubeconfig
                                                                   kubectl get apiservice
NAME                                 SERVICE                     AVAILABLE   AGE
v0alpha1.example.grafana.app         grafana/example-apiserver   True        3m42s
v0alpha1.featuretoggle.grafana.app   Local                       True        73m
v0alpha1.playlist.grafana.app        Local                       True        73m
v0alpha1.service.grafana.app         Local                       True        73m

todd@work ~/g/s/g/g/grafana (toddtreece/standalone-aggregation) [127]> kubectl api-resources
NAME             SHORTNAMES   APIVERSION                           NAMESPACED   KIND
apiservices                   apiregistration.k8s.io/v1            false        APIService
dummy                         example.grafana.app/v0alpha1         true         DummyResource
runtime                       example.grafana.app/v0alpha1         false        RuntimeInfo
features                      featuretoggle.grafana.app/v0alpha1   false        Feature
featuretoggles                featuretoggle.grafana.app/v0alpha1   true         FeatureToggles
playlists                     playlist.grafana.app/v0alpha1        true         Playlist
externalnames                 service.grafana.app/v0alpha1         true         ExternalName
```

File: pkg/services/apiserver/aggregator/README.md [hunk 7]
```diff
@@ -45,3 +45,52 @@ live under that instead.
 `service.grafana.app`) and `data/grafana-apiextensions` (`apiextensions.k8s.io`).
 2. Since `grafana-aggregator` outputs configuration (TLS and kubeconfig) that is used in the invocation of aggregated
   servers, ensure you start the aggregated service after launching the aggregator during local development.
+
+## How to get started
```
Comment (toddtreece) [inline pkg/services/apiserver/aggregator/README.md:49]: @ryantxu i added docs and an example based on what @charandas wrote for the standalone aggregator command. remote seems to work as expected:

```shell
export KUBECONFIG=$PWD/data/grafana-apiserver/grafana.kubeconfig
kubectl get apiservices
NAME                                 SERVICE                     AVAILABLE   AGE
v0alpha1.example.grafana.app         grafana/example-apiserver   True        3m42s
v0alpha1.featuretoggle.grafana.app   Local                       True        73m
v0alpha1.playlist.grafana.app        Local                       True        73m
v0alpha1.service.grafana.app         Local                       True        73m
```

```shell
todd@work ~/g/s/g/g/grafana (toddtreece/standalone-aggregation) [127]> kubectl api-resources
NAME             SHORTNAMES   APIVERSION                           NAMESPACED   KIND
apiservices                   apiregistration.k8s.io/v1            false        APIService
dummy                         example.grafana.app/v0alpha1         true         DummyResource
runtime                       example.grafana.app/v0alpha1         false        RuntimeInfo
features                      featuretoggle.grafana.app/v0alpha1   false        Feature
featuretoggles                featuretoggle.grafana.app/v0alpha1   true         FeatureToggles
playlists                     playlist.grafana.app/v0alpha1        true         Playlist
externalnames                 service.grafana.app/v0alpha1         true         ExternalName
```

File: pkg/services/apiserver/auth/authorizer/provider.go (hunk 6 above)
Comment (toddtreece) [reply pkg/services/apiserver/auth/authorizer/provider.go:25]: i think it's only possible via the loopback config

File: pkg/services/apiserver/aggregator/aggregator.go [hunk 8]
```diff
@@ -198,16 +196,20 @@ func makeAPIServiceAvailableHealthCheck(name string, apiServices []*v1.APIServic
 	}
 
 	// Watch add/update events for APIServices
-	_, _ = apiServiceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+	_, err := apiServiceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc:    func(obj interface{}) { handleAPIServiceChange(obj.(*v1.APIService)) },
 		UpdateFunc: func(old, new interface{}) { handleAPIServiceChange(new.(*v1.APIService)) },
 	})
```
Comment (radiohead) [inline pkg/services/apiserver/aggregator/aggregator.go:202]: @toddtreece what happens if an APIService is added, never becomes ready and then gets removed?

IIUC this means that the healthcheck will never succeed because we will not remove the that service from the list of pending APIServices. If that's correct, we may want to have a delete handler here as well.

Comment (toddtreece) [reply pkg/services/apiserver/aggregator/aggregator.go:202]: @radiohead good question. this was copied from [kube-apiserver](https://github.com/kubernetes/kubernetes/blob/73f19e4c0162a2ee711c254b2af226b4e135d0e3/cmd/kube-apiserver/app/aggregator.go#L233), and the list of pending services used for the health check is only for the API groups that are registered on startup, which for `kube-apiserver` are groups like the core v1 API groups. services registered on startup are all local delegation, and bypass any ready checks and are considered always available. i think unless something is very broken right now it's probably not going to be an issue because the API group should be removed from the pending list during the startup process. @charandas this is a great observation, and something we should keep in mind if we end up needing to inject remote services into the list of startup groups.

**Review by charandas — APPROVED** [review]
Added some changes to README. Other than that, LGTM :shipit:
